### ISSUE TITLE
in_app_purchase can't dispatch purchase updates.

### ISSUE THREAD

**Issue by Froelund** [flutter/flutter#34040]
After upgrading to `0.1.1+2` of the in_app_purchase plugin, we're unable to get purchase updates on Android.

I've been digging through the stack, and I think I have located the issue. When the `BillingClientWrapper#callHandler` calls the `PurchasesUpdatedListener` it tries to pass the `call.arguments`  to `PurchasesResultWrapper#fromJson` factory.

In commit [0dcd0cd931337a4759daee2796c18c1dd7df4dcd](https://github.com/flutter/plugins/commit/0dcd0cd931337a4759daee2796c18c1dd7df4dcd) a new signature was added to the factory, which makes the factory call fail: 
`factory PurchasesResultWrapper.fromJson(Map map)`
was changed to
`factory SkProductResponseWrapper.fromJson(Map<String, dynamic> map)`

The consequence of this, is that the `PurchasesUpdatedListener#onPurchasesUpdated(int, List<Purchase>)` is simply dropped.

**Comment by Froelund** [comment]
I'm quite certain that this means that no-one will be able to complete purchases on flutter `stable` channel using in_app_purchase version `0.1.1+2`.
I came around this issue by sticking to `0.1.1+1`.

**Comment by rogiervandenberg** [comment]
I can confirm that, I receive an `W/BillingHelper(26580): Couldn't find purchase lists, trying to find single data.` error after a successful purchase, but my InAppPurchaseConnection.instance.purchaseUpdatedStream is not updated. Only after reloading and asking for pastPurchases, I receive an event on my stream. Solution was to revert from `0.1.1+2` to `0.1.1+1`

**Comment by rahulraj64** [comment]
I am building from the source and also experiencing the same problem still in 0.2.0. This is a critical issue.

**Comment by rickdroio** [comment]
is there any update?

**Comment by Froelund** [comment]
I just checked pub.dev, and it seems that two new versions has been published. I haven't tested those myself. I do see that @cyanglaz has been committing to the plugin since this issue has been created. This issue does seem critical :|

**Comment by cyanglaz** [comment]
The fix is here https://github.com/flutter/plugins/pull/1775
Will close the issue when the PR is merged

**Comment by cyanglaz** [comment]
Fix is merged, closing the issue. If you found related bugs please feel free to file another issue with details.

**Comment by rahulraj64** [comment]
Can anyone verify that the issue is resolved? I am still experiencing the same issue even in 0.2.0+1
To reproduce, simply purchase a subscription from Android, the _purchaseUpdatedStream_ is not updated. It is working only in 0.1.1+1

**Comment by pabloserranof** [comment]
I am also experiencing the same issue with 0.2.0+1. The _purchaseUpdatedStream_ is not updated after a purchase.

**Comment by cyanglaz** [comment]
Can anyone provide a mini sample app to reproduce the issue. It seems working for me with the example app in the plugin.

**Comment by pabloserranof** [comment]
Getting the same error reported in this issue also with the example app in the plugin with 0.2.0+1.
I tried with the beta and stable channel with the same result.
The _purchaseUpdatedStream_ is not updated after a purchase. A relaunch of the app shows the subscription as purchased. Only tried on Android. 
Please let me know if any other info may be useful. Thanks!


```
flutter doctor -v

[✓] Flutter (Channel stable, v1.5.4-hotfix.2, on Mac OS X 10.14.5 18F132, locale es-ES)
    • Flutter version 1.5.4-hotfix.2 at /Users/developer/flutter
    • Framework revision 7a4c33425d (9 weeks ago), 2019-04-29 11:05:24 -0700
    • Engine revision 52c7a1e849
    • Dart version 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/developer/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.0, Build version 10A255
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 36.1.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] Connected device (1 available)
    • SM G960F • 2168360e37047ece • android-arm64 • Android 9 (API 28)

• No issues found!

```

**Comment by rahulraj64** [comment]
The same example app is not working for me too. I cloned the plugin repo and build from the source. I made sure the upstream branch is up to date with the latest commit _[in_app_purchase]fix the type casting issue (#1775)_ and the latest version is _0.2.0+1_. When purchasing a subscription from Android, purchaseUpdatedStream is not updated. I checked with the master and stable branches multiple times. 
@cyanglaz, This can be easily reproduced by manually test a **subscription purchase** from **Android** with the example app itself.

**Comment by cyanglaz** [comment]
@rahulraj64 @pabloserranof  Thank you for the details, did you mean only subscription causes the issue? Consumable works fine?

**Comment by rahulraj64** [comment]
@cyanglaz It's not working for managed products either in 0.2.0+1. But when I used the version  0.1.1+1 from the pub, it worked. Tested on Samsung Galaxy S8 running Android 9.0. Please feel free to ask for any additional details from our side. I am desperately waiting for integrating this plugin into my production app.

**Comment by pabloserranof** [comment]
@cyanglaz I just tried with a managed product for the first time with 0.2.0+1 and it does not work either. Getting the same result as with the subscription reported before. Tested with the example app code from the plugin and with the same "flutter doctor" as before. The item is purchased but _purchaseUpdatedStream_ is not updated.

**Comment by radrig** [comment]
I have encountered the same problem with 0.1.1 and 0.2.0 as well. Always get "Couldn't find purchase lists, trying to find single data."  Just wondering if there is going to be updates coming soon or if I should explore other avenues.

**Comment by sageata** [comment]
I did a bit of digging and it seems that the function that sets up the stream never gets triggered.
I placed break points and also print statements in the plugin function that is meant to return the StreamSubscription but it never gets called. 

I hope this helps.

Let me know if this is not clear and you need more details.

**Comment by cyanglaz** [comment]
Thank you for all the valuable details. I have detected the bug and will provide a fix soon.

**Comment by cyanglaz** [comment]
https://github.com/flutter/plugins/pull/1785 Should fix the issue in 0.2.0+1, if you update to 0.2.0+2, the stream should be able to get updates. Closing the issue now. If related issues are still found, please feel free to open the issue or ping me in this issue.

**Comment by rahulraj64** [comment]
Verified that the fix 0.2.0+2 is working. 
Thanks @cyanglaz

**Comment by HiroshiHorie** [comment]
Is this fixed? I am still having issues with the latest version.

**Comment by cyanglaz** [comment]
@HiroshiHorie This is fixed. You might have a different issue. Do you mind open another issue ticket explaining the details of the issue that you are facing?

**Comment by HiroshiHorie** [comment]
@cyanglaz I'm sorry it seems to be fixed, I don't know why there was a time it didn't work.

**Comment by solokingjec** [comment]
This issue is still present on iOS. Android seems perfect

**Comment by Froelund** [comment]
@solokingjec If that's the case, I think you should create a new issue with a reference to this one.

**Comment by pawlowskim** [comment]
I was able to get previous not finished purchases when I start listener in ```main()```, but before ```runApp()```. When it was started before main, or somewhere else later on some page it didn't return previous purchases. New ones arrived correctly (even if they were not finished, I received message that it's already bought and will be restored but they came through listener).

**Comment by ollydixon** [comment]
I'm seeing this issue again on ```0.3.4```...

**Comment by kepuza** [comment]
purchaseUpdated.listen not working. I'm experimenting with test card with rejection after a few minutes but its not trigger

**Comment by ollydixon** [comment]
@kepuza the maintainers of this lib have never been able to fix this lib in over two years. Don't bother with it; we've been using https://pub.dev/packages/flutter_inapp_purchase and it works flawlessly (for two years)

**Comment by kepuza** [comment]
@ollydixon Thank you i will try
